pkg/monitors/pingdomtransaction/pingdom-transaction-monitor.go

```diff
@@ -0,0 +1,386 @@
+
+	monitors := service.GetAll()
+	for _, mon := range monitors {
+		if mon.Name == name {
+			return &mon, nil
+		}
+	}
+
+	return match, fmt.Errorf("unable to locate monitor with name %v", name)
+}
+
+func (service *PingdomTransactionMonitorService) GetAll() []models.Monitor {
+	var monitors []models.Monitor
+	checks, _, err := service.client.TMSChecksAPI.GetAllChecks(service.context).Type_("script").Execute()
+	if err != nil {
+		log.Error(err, "Error getting all transaction checks")
+		return monitors
+	}
+
+	if checks == nil {
+		return monitors
+	}
+	for _, mon := range checks.GetChecks() {
+		newMon := models.Monitor{
+			URL:  service.GetUrlFromSteps(*mon.Id),
+			ID:   fmt.Sprintf("%v", *mon.Id),
+			Name: *mon.Name,
+		}
+		monitors = append(monitors, newMon)
+	}
+	return monitors
+}
+
+func (service *PingdomTransactionMonitorService) GetUrlFromSteps(id int64) string {
+	check, _, err := service.client.TMSChecksAPI.GetCheck(service.context, id).Execute()
+	if err != nil {
+		log.Error(err, "Error getting transaction check")
+		return ""
+	}
+	if check == nil {
+		return ""
+	}
+	for _, step := range check.GetSteps() {
+		if step.GetFn() == "go_to" {
+			return *step.GetArgs().Url
+		}
+	}
+	return ""
+}
+
+func (service *PingdomTransactionMonitorService) Add(m models.Monitor) {
+	transactionCheck := service.createTransactionCheck(m)
+	if transactionCheck == nil {
+		return
+	}
+	_, resp, err := service.client.TMSChecksAPI.AddCheck(service.context).CheckWithoutID(*transactionCheck).Execute()
+	if err != nil {
+		log.Error(err, "Error Adding Pingdom Transaction Monitor "+m.Name, "Response", parseResponseBody(resp))
+	} else {
+		log.Info("Successfully added Pingdom Transaction Monitor " + m.Name)
+	}
+}
+
+func (service *PingdomTransactionMonitorService) Update(m models.Monitor) {
+	transactionCheck := service.createTransactionCheck(m)
+	if transactionCheck == nil {
+		return
+	}
+	monitorID := util.StrToInt64(m.ID)
+	_, resp, err := service.client.TMSChecksAPI.ModifyCheck(service.context, monitorID).CheckWithoutIDPUT(*transactionCheck.AsPut()).Execute()
+	if err != nil {
+		log.Error(err, "Error Updating Pingdom Transaction Monitor", "Response", parseResponseBody(resp))
+		return
+	}
+	log.Info("Updated Pingdom Transaction Monitor Monitor " + m.Name)
+}
+
+func (service *PingdomTransactionMonitorService) Remove(m models.Monitor) {
+	_, resp, err := service.client.TMSChecksAPI.DeleteCheck(service.context, util.StrToInt64(m.ID)).Execute()
+	if err != nil {
+		log.Error(err, "Error Deleting Pingdom Transaction Monitor", "Response", parseResponseBody(resp))
+	} else {
+		log.Info("Deleted Pingdom Transaction Monitor Monitor " + m.Name)
+	}
+}
+
+func (service *PingdomTransactionMonitorService) createTransactionCheck(monitor models.Monitor) *pingdomNew.CheckWithoutID {
+	transactionCheck := &pingdomNew.CheckWithoutID{}
+	providerConfig, _ := monitor.Config.(*endpointmonitorv1alpha1.PingdomTransactionConfig)
+	if providerConfig == nil {
```

Comment:
shouldnt this be checked at setup level and not so down the code flow?

Reply:
the providerConfig is set on alert configuration. We can not move to setup because it needs to be checked for every alert. One alternative instead of checking for existence of `*endpointmonitorv1alpha1.PingdomTransactionConfig` we can use the providers from the spec https://github.com/stakater/IngressMonitorController/blob/f62cd9b32c37a54bed40a6c929214655225dfb06/api/v1alpha1/endpointmonitor_types.go#L37